src/server/transaction.cc

```diff
@@ -1052,34 +1044,46 @@ bool Transaction::ScheduleInShard(EngineShard* shard, bool execute_optimistic) {
   if (txid_ > 0 && shard->committed_txid() >= txid_)
     return false;
 
+  auto acquire_fp_locks = [&](bool shard_unlocked) mutable {
+    bool keys_unlocked = GetDbSlice(shard->shard_id()).Acquire(mode, lock_args);
+    lock_granted = shard_unlocked && keys_unlocked;
+
+    sd.local_mask |= KEYLOCK_ACQUIRED;
+    if (lock_granted) {
+      sd.local_mask |= OUT_OF_ORDER;
+    }
+
+    DVLOG(3) << "Lock granted " << lock_granted << " for trans " << DebugId();
+  };
+
+  auto release_fp_locks = [&]() {
+    GetDbSlice(shard->shard_id()).Release(mode, lock_args);
+    sd.local_mask &= ~KEYLOCK_ACQUIRED;
+  };
+
   // Acquire intent locks. Intent locks are always acquired, even if already locked by others.
   if (!IsGlobal()) {
     lock_args = GetLockArgs(shard->shard_id());
     bool shard_unlocked = shard->shard_lock()->Check(mode);
 
+    // We need to acquire the fp locks because the executing callback
+    // within RunCallback below might preempt.
+    acquire_fp_locks(shard_unlocked);
```

Comment:
nits:
1. Maybe we don't need acquire_fp_locks as a separate function
2. If you decide to keep it, maybe return lock_granted, otherwise the flow might be somewhat confusing, looking at the if below (i.e. not clear where lock_granted changed)